fabric-common-tags-api-v1/src/datagen/java/net/fabricmc/fabric/impl/tag/common/datagen/generators/BiomeTagGenerator.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.tag.common.datagen.generators;
+
+import net.minecraft.tag.BiomeTags;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.BiomeKeys;
+
+import net.fabricmc.fabric.api.datagen.v1.FabricDataGenerator;
+import net.fabricmc.fabric.api.datagen.v1.provider.FabricTagProvider;
+import net.fabricmc.fabric.api.tag.v1.CommonBiomeTags;
+
+public class BiomeTagGenerator extends FabricTagProvider.DynamicRegistryTagProvider<Biome> {
+	public BiomeTagGenerator(FabricDataGenerator dataGenerator) {
+		super(dataGenerator, Registry.BIOME_KEY, "biomes", "Biome Tags");
+	}
+
+	@Override
+	protected void generateTags() {
+		generateDimensionTags();
+		generateCategoryTags();
+		generateOtherBiomeTypes();
+		generateClimateAndVegetationTags();
+		generateTerrainDescriptorTags();
+	}
+
+	private void generateDimensionTags() {
+		getOrCreateTagBuilder(CommonBiomeTags.IN_NETHER)
+				.addOptionalTag(BiomeTags.IS_NETHER)
+				.add(BiomeKeys.CRIMSON_FOREST)
+				.add(BiomeKeys.WARPED_FOREST)
+				.add(BiomeKeys.NETHER_WASTES)
+				.add(BiomeKeys.SOUL_SAND_VALLEY)
+				.add(BiomeKeys.BASALT_DELTAS);
+		getOrCreateTagBuilder(CommonBiomeTags.IN_THE_END)
+				.add(BiomeKeys.END_BARRENS)
+				.add(BiomeKeys.END_MIDLANDS)
+				.add(BiomeKeys.END_HIGHLANDS)
+				.add(BiomeKeys.THE_END)
+				.add(BiomeKeys.SMALL_END_ISLANDS)
+				.add(BiomeKeys.THE_VOID);
+		// We avoid the vanilla group tags here as mods may add to them without actually spawning them in the overworld
+		getOrCreateTagBuilder(CommonBiomeTags.IN_OVERWORLD)
+				.addOptionalTag(BiomeTags.STRONGHOLD_HAS_STRUCTURE)// This has most overworld biomes already
```

Comment:
Data packs may want to modify the biomes that strongholds can spawn in without removing those biomes from the overworld altogether. Using the `minecraft:has_structure/stronghold` tag in the `c:in_overworld` tag breaks that assumption.